openedx/core/djangoapps/oauth_dispatch/docs/decisions/0009-jwt-in-session-cookie.rst

```diff
@@ -0,0 +1,193 @@
+   newly logged in user. The JWT will be embedded in a session cookie ("JWT Cookie"), included in the login
+   response and stored in the user's browser cookie jar.
+
+#. **Automatically extract JWT from Cookie on API calls.** The `Django Rest Framework JWT`_ library we use supports a
+   JWT_AUTH_COOKIE_ configuration setting to specify the JWT cookie name. When set, the JSONWebTokenAuthentication_
+   class `automatically extracts the JWT from the cookie`_. Since all open edX REST endpoints that support JWT-based
+   authentication derive from this base class, microfrontends can access them with JWT Cookies. We will enable this
+   setting.
+
+.. _Django Rest Framework JWT: https://getblimp.github.io/django-rest-framework-jwt/
+.. _JWT_AUTH_COOKIE: https://github.com/GetBlimp/django-rest-framework-jwt/blob/master/docs/index.md#jwt_auth_cookie
+.. _JSONWebTokenAuthentication: https://github.com/GetBlimp/django-rest-framework-jwt/blob/0a0bd402ec21fd6b9a5f715d114411836fbb2923/rest_framework_jwt/authentication.py#L71
+.. _automatically extracts the JWT from the cookie: https://github.com/GetBlimp/django-rest-framework-jwt/blob/0a0bd402ec21fd6b9a5f715d114411836fbb2923/rest_framework_jwt/authentication.py#L86-L87
+
+
+JWT Cookie Lifetime
+^^^^^^^^^^^^^^^^^^^
+
+#. **Cookie and JWT expiration.** Both the session cookie and the JWT have expiration times.
+
+   * For simplicity and consistency, the session cookie and its containing JWT will expire at the same time. There's
+     no need to have these be different values.
+
+   * Given this, JWT cookies will always have expiration values, unlike `current open edX session cookies that may
+     have no expiration`_.
+
+   * A configuration setting, JWT_AUTH_COOKIE_EXPIRATION, will specify the expiration duration for JWTs and their
+     containing cookie.
+
+#. **Revocation with short-lived JWTs** Given the tradeoff between long-lived JWTs versus immediacy of revocation, we
+   need to configure an appropriate expiration value for JWT cookies. In a future world with an API gateway, we can
+   have longer lived JWTs with a stateful check against a centralized `JWT blacklist`_ and each JWT uniquely
+   identified by a `JWT ID (jti)`_. In the meantime, we will err on the side of security and have short-lived JWTs. 
+
+#. **Refresh JWT Cookies.** When a JWT expires, we do not want to ask the user to login again while their browser
+   session remains alive. A microfrontend will detect JWT expiration upon receiving a 401 response from an API
+   endpoint. To automatically refresh the JWT cookie, the microfrontend will call a new endpoint ("refresh") that
+   returns a new JWT Cookie to keep the user's session alive.
+
+   * To support this, the login endpoint will include 2 related cookies in its response:
+
+     * **JWT Cookie** (as described above), with a *domain* setting so that it is forwarded to any microservice in
```

Comment:
Does this whole JWT cookie architecture support microservices that are on different domains? We definitely have a use case where we want a microfrontend and microbackend on a non-edx.org domain name to have SSO with the courses.edx.org LMS and to be able to make API calls on behalf of the logged-in edX user, and I'm sure other Open edX instances would have similar requirements.

Reply:
I think for that use case we would employ an OAuth2 flow using the authorization code grant type. Your microbackend would be registered as a DOT application in LMS.

JWT cookies would be used for SSO across microservices that are served from same domain or set of sub-domains, e.g. openedxsite.org, lms.openedxsite.org, ecommerce.openedxsite.org, etc.